server/session/handler_enchanting.go

```diff
@@ -0,0 +1,341 @@
+package session
+
+import (
+	"fmt"
+	"github.com/df-mc/dragonfly/server/block"
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/internal/sliceutil"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/sandertv/gophertunnel/minecraft/protocol"
+	"github.com/sandertv/gophertunnel/minecraft/protocol/packet"
+	"golang.org/x/exp/slices"
+	"math"
+	"math/rand"
+)
+
+// enchantNames are names translated to the 'Standard Galactic Alphabet' client-side. The names generally have no meaning
+// on the vanilla server implementation, so we can sneak some easter eggs in here without anyone noticing.
+var enchantNames = []string{
+	"dragonfly",
+	"gophertunnel",
+	"go raknet",
+	"go lang",
+	"sandertv",
+	"t 14 raptor",
+	"da pig guy",
+	"potatoe train yt",
+}
+
+const (
+	// enchantingInputSlot is the slot index of the input item in the enchanting table.
+	enchantingInputSlot = 0x0E
+	// enchantingLapisSlot is the slot index of the lapis in the enchanting table.
+	enchantingLapisSlot = 0x0F
+)
+
+// handleEnchant handles the enchantment of an item using the CraftRecipe stack request action.
+func (h *ItemStackRequestHandler) handleEnchant(a *protocol.CraftRecipeStackRequestAction, s *Session) error {
+	// First ensure that the selected slot is not out of bounds.
+	if a.RecipeNetworkID > 2 {
+		return fmt.Errorf("invalid recipe network id: %d", a.RecipeNetworkID)
+	}
+
+	// Now ensure we have an input and only one input.
+	input, err := s.ui.Item(enchantingInputSlot)
+	if err != nil {
+		return err
+	}
+	if input.Count() > 1 {
+		return fmt.Errorf("enchanting tables only accept one item at a time")
+	}
+
+	// Determine the available enchantments using the session's enchantment seed.
+	allCosts, allEnchants := s.determineAvailableEnchantments(s.c.World(), s.openedPos.Load(), input)
+	if len(allEnchants) == 0 {
+		return fmt.Errorf("can't enchant non-enchantable item")
+	}
+
+	// Use the slot plus one as the cost. The requirement and enchantments can be found in the results from
+	// determineAvailableEnchantments using the same slot index.
+	cost := int(a.RecipeNetworkID + 1)
+	requirement := allCosts[a.RecipeNetworkID]
+	enchants := allEnchants[a.RecipeNetworkID]
+
+	// If we don't have indefinite resources, we need to deduct Lapis Lazuli and experience.
```

Comment:
Would call this `infinite` as opposed to `indefinite`, because that's more indicative of a time span.

Reply:
Renamed.